### ISSUE TITLE
MAP4 implementation is wrong

### ISSUE THREAD

**Issue by LucaCappelletti94** [scikit-fingerprints/scikit-fingerprints#303]
Hi!

By chance, I believe I have found definitive proof that the [current implementation of MAP4 available in this library](https://github.com/scikit-fingerprints/scikit-fingerprints/blob/05326d0504ac9c18cc7e5924404ea12b2e280016/skfp/fingerprints/map.py#L21-L141) must contain errors.

My claim is based on two pieces of evidence: visualizations and model performance. Since the [original code](https://github.com/reymond-group/map4) is admittedly hard to use, [I have fully refactored it myself here](https://github.com/LucaCappelletti94/map4) and [published it on pip](https://pypi.org/project/map4/). I believe my implementation to pedissequally follow their original code and paper, minus some dumb points in their code (they were converting things to a set then to a list then to a set and other such things). I [have opened a pull request in their original repo](https://github.com/reymond-group/map4/pull/27), but they do not seem interested in replying.

By having a different implementation, I was able to compare your implementation with mine.

Both visualizations and performance report to the tasks I am working on, namely the prediction of chemical taxonomy given as input for the fingerprint. The names of the chemical taxonomy layers are "pathways", "superclasses" and "classes". Since chemicals may have multiple labels, the task is defined as a multi-label task.

## First evidence: Feature Visualizations

Here are the PCA and TSNE decompositions of the MAP implementation available from scikit-fingerprints, and [map4](https://github.com/LucaCappelletti94/map4), respectively. The colors are not some model predictions, but the ground truth labelling associated with this dataset ([dataset from this paper](https://pubs.acs.org/doi/10.1021/acs.jnatprod.1c00399)). While there is admittedly some uncertainty on the quality of some of these labels, I expect most of the labels to be correct. For each chemical taxonomical layer, the most common label is shown.

![Scikit-fingerprint's variant](https://github.com/LucaCappelletti94/hammer/blob/main/data_visualizations/MinHashed%20Atom%20Pair%20(2r,%202048b).png?raw=true)

![MAP4](https://github.com/LucaCappelletti94/hammer/blob/main/data_visualizations/MAP4.png?raw=true)

You can find several other such visualizations for other fingerprints made available from scikit-fingerprint [here](https://github.com/LucaCappelletti94/hammer/tree/main/data_visualizations). They all work nominally, with the singular exception of `MAP` which shows what I can describe by my experience as a near-total decomposition failure, both from the PCA and TSNE perspective.

## Second evidence: Performance

In this second plot, you can find the AUPRC performance of several fingerprints provided by the [scikit-fingerprint](https://github.com/scikit-fingerprints/scikit-fingerprints) package. You can [find more metrics here](https://github.com/LucaCappelletti94/hammer/tree/main/feature_sets_evaluation_barplots). The difference in performance between MAP4 (in blue) and MinHashed Atom Pair (in green), averaged across 5 holdouts and using the same parameters, is extreme. The model used in this case is a tensorflow neural network, but I have obtained analogous results using a sklearn Random Forest.

![Performance](https://github.com/LucaCappelletti94/hammer/blob/main/feature_sets_evaluation_barplots/class_auprc_feature_sets.png?raw=true)

## My proposition

To resolve this issue, since it looks like my refactored implementation works well, I would suggest we swap the current `MAP` re-implementation present in this package with an interface to my implementation. Happy to change anything you believe to be needed in my implementation to make it more compatible with this package. As it stands, I tried to leave it unchanged in respect to the original one, but they don't seem to care much for my PR.

If you'd like for me to proceed with a PR with this swap, let me know.

Luca

**Comment by j-adamczyk** [comment]
@LucaCappelletti94 this is a great idea, and congrats on improving the original implementation. The bad results also surprised me during our experiments, so I would very much appreciate a fix here.

The main decision here, I think, is whether to use `map4` as a separate package, or to include the whole code into `scikit-fingerprints` (BTW, very small detail, but note it's plural "fingerprints"). A separate package may be useful if one needs only that fingerprint, but this is basically a single file implementation, so in my opinion including it fully in `scikit-fingerprints` would be reasonable. It would also make dependency management easier. Of course, this is your decision here.

One concern I have is [using the MHFPEncoder here](https://github.com/LucaCappelletti94/map4/blob/master/map4/map4.py#L54). It requires installing `mfhp`, which [is maintained by the same group as MAP4](https://github.com/reymond-group/mhfp/tree/master). There is already an issue there related to NumPy 2.0 compatibility, so I would like to avoid that. Fortunately, the [MHFPEncoder implementation](https://github.com/reymond-group/mhfp/blob/master/mhfp/encoder.py) also looks quite easy, and not a lot of code, it's mostly comments. They also use the `hashlib` for the actual hashing - this would enable us to remove the `datasketch` dependency from `scikit-fingerprints`. By a quick glance, it could also be improved a bit, like their MAP4 implementation.

So my general suggestion is:
- also reimplement `MHFPEncoder`
- use your MAP4 implementation
- include this in `scikit-fingerprints`

Do you agree?

**Comment by LucaCappelletti94** [comment]
I mostly refactored MAP4 into a standalone package as I had not encountered yet this one, so sure I can make a PR to add it here. If the original authors were a bit more reactive on GitHub I'd have preferred to try and keep the original implementation working, to have a valid "ground truth". The next best thing certainly is to have it all here without odd dependencies, so yes I can add my refactored MAP4 and refactor MHFPEncoder.

The only issue is that I am still having issues running the test suite coverage at this time, so while I can add these implementations and a reasonable test suite, I will be blind regarding the test coverage.

Before proceeding with this PR though, I'd like to finalize the TQDM PR thing, so as not to have different PR edits in potential collision.

**Comment by j-adamczyk** [comment]
Sure. The test coverage one is weird, as I could not reproduce this on a fresh `scikit-fingerprints` installation. Python 3.9 + new Poetry environment + `make setup` worked with no problem. I tested this on both Ubuntu 22.04 and Windows 10. But generally speaking we're not that interested in percentage of coverage, just running PyTest is good enough, since test coverage clearly does not pick up all checked code (e.g. count MACCS).

**Comment by LucaCappelletti94** [comment]
Ok, then let's finish this TQDM thing and I'll pick up this other thing when we are done.
